ddtrace/tracer/textmap.go

```diff
@@ -462,3 +475,137 @@ func (*propagatorB3) extractTextMap(reader TextMapReader) (ddtrace.SpanContext,
 	}
 	return &ctx, nil
 }
+
+const (
+	traceparentHeader = "traceparent"
+	tracestateHeader  = "tracestate"
+)
+
+// propagatorW3c implements Propagator and injects/extracts span contexts
+// using W3C tracecontext/traceparent headers. Only TextMap carriers are supported.
+type propagatorW3c struct{}
+
+func (p *propagatorW3c) Inject(context ddtrace.SpanContext, carrier interface{}) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (p *propagatorW3c) Extract(carrier interface{}) (ddtrace.SpanContext, error) {
+	switch c := carrier.(type) {
+	case TextMapReader:
+		return p.extractTextMap(c)
+	default:
+		return nil, ErrInvalidCarrier
+	}
+}
+
+func (*propagatorW3c) extractTextMap(reader TextMapReader) (ddtrace.SpanContext, error) {
+	var parentHeader string
+	stateHeaders := []string{}
+	// to avoid parsing tracestate header(s) if traceparent is invalid
+	err := reader.ForeachKey(func(k, v string) error {
+		key := strings.ToLower(k)
+		switch key {
+		case traceparentHeader:
+			if parentHeader != "" {
+				return ErrSpanContextCorrupted
+			}
+			parentHeader = v
+		case tracestateHeader:
+			stateHeaders = append(stateHeaders, v)
+		}
+		return nil
+	})
+	if err != nil || parentHeader == "" {
+		return nil, ErrSpanContextCorrupted
+	}
+	var ctx spanContext
+	err = parseTraceparent(&ctx, parentHeader)
+	if err != nil {
+		return nil, ErrSpanContextNotFound
+	}
+	err = parseTracestate(&ctx, stateHeaders)
+	if ctx.traceID == 0 || ctx.spanID == 0 {
+		return nil, ErrSpanContextNotFound
+	}
+	return &ctx, nil
+}
+
+func parseTraceparent(ctx *spanContext, v string) error {
+	v = strings.Trim(v, "\t -")
+	if len(v) != 55 {
+		return ErrSpanContextCorrupted
+	}
+	version, err := strconv.ParseUint(v[:2], 16, 64)
```

Comment:
I wonder if we could use https://pkg.go.dev/fmt#Sscanf for parsing the digits out of the string. Worth playing around with.